### PR TITLE
[Enhancement][DLA] IcebergScanNode do not send correct min max conjouncts to be

### PR DESCRIPTION
**Pull request by caneGuy**
## What type of PR is this：
- [x] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

This pr fix min/max predicates missed for IcebergScanNode, it will improve some query perfomance
## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 4 / 41 (09.76%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/IcebergScanImplementationRule.java|0|1|00.00%|[31]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|0|4|00.00%|[734, 776, 832, 833]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerTraceUtil.java|0|2|00.00%|[193, 426]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalIcebergScanOperator.java|0|17|00.00%|[22, 31, 36, 41, 66, 67, 69, 70, 71, 72, 73, 74, 79, 85, 86, 87, 88]|
|:large_blue_circle:|com/starrocks/planner/IcebergScanNode.java|0|10|00.00%|[50, 136, 137, 190, 214, 216, 244, 246, 289, 295]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalIcebergScanOperator.java|3|6|50.00%|[49, 74, 86]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneHDFSScanColumnRule.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 5 / 49 (10.20%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/IcebergScanImplementationRule.java|0|1|00.00%|[31]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|0|4|00.00%|[734, 776, 832, 833]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerTraceUtil.java|0|2|00.00%|[193, 426]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalIcebergScanOperator.java|0|17|00.00%|[22, 31, 36, 41, 66, 67, 69, 70, 71, 72, 73, 74, 79, 85, 86, 87, 88]|
|:large_blue_circle:|com/starrocks/planner/IcebergScanNode.java|0|10|00.00%|[50, 136, 137, 190, 214, 216, 244, 246, 289, 295]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RemoteScanPartitionPruneRule.java|1|8|12.50%|[68, 70, 71, 72, 74, 75, 76]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalIcebergScanOperator.java|3|6|50.00%|[49, 74, 86]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneHDFSScanColumnRule.java|1|1|100.00%|[]|

**Comment by caneGuy** [comment]
image

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/physical/PhysicalIcebergScanOperator.java [hunk 1]
```diff
@@ -72,24 +63,29 @@ public boolean equals(Object o) {
...
         }
 
-        PhysicalIcebergScanOperator that = (PhysicalIcebergScanOperator) o;
+        PhysicalHiveScanOperator that = (PhysicalHiveScanOperator) o;
+        ScanOperatorPredicates targetPredicts = ((PhysicalHiveScanOperator) o).getScanOperatorPredicates();
         return Objects.equal(table, that.table) &&
-                Objects.equal(conjuncts, that.conjuncts) &&
-                Objects.equal(minMaxConjuncts, that.minMaxConjuncts) &&
-                Objects.equal(minMaxColumnRefMap, that.minMaxColumnRefMap);
+                Objects.equal(predicates.getSelectedPartitionIds(), targetPredicts.getSelectedPartitionIds()) &&
```
Comment (dirtysalt) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/physical/PhysicalIcebergScanOperator.java:69]: Do we have equal method in `ScanOperatorPredicates`?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoteScanPartitionPruneRule.java [hunk 2]
```diff
@@ -62,9 +65,16 @@ public List<OptExpression> transform(OptExpression input, OptimizerContext conte
         Map<ColumnRefOperator, Set<Long>> columnToNullPartitions = Maps.newHashMap();
 
         try {
-            initPartitionInfo(operator, columnToPartitionValuesMap, columnToNullPartitions, context);
-            classifyConjuncts(operator, columnToPartitionValuesMap);
-            computePartitionInfo(operator, columnToPartitionValuesMap, columnToNullPartitions, context);
+            if (operator.getTable() instanceof IcebergTable) {
+                // A little trick to set min max for iceberg table here
+                classifyConjuncts(operator, columnToPartitionValuesMap);
```
Comment (dirtysalt) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoteScanPartitionPruneRule.java:70]: Could this piece of code be integrated into the following framework? So there is no if/else stmt.

```
initPartitionInfo(operator, columnToPartitionValuesMap, columnToNullPartitions, context);
                classifyConjuncts(operator, columnToPartitionValuesMap);
                computePartitionInfo(operator, columnToPartitionValuesMap, columnToNullPartitions, context);
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoteScanPartitionPruneRule.java [hunk 3]
```diff
@@ -62,9 +65,16 @@ public List<OptExpression> transform(OptExpression input, OptimizerContext conte
...
+                LOG.info("Add min max conjuncts for iceberg table!" + ((IcebergTable) operator.getTable()).getTable());
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RemoteScanPartitionPruneRule.java:71]: remove or change to debug level

**Comment by DorianZheng** [comment]
Is it better to separate another rule to compute min/max conjuncts?

**Comment by caneGuy** [comment]
> Is it better to separate another rule to compute min/max conjuncts?

actually i think we should do some refactor for this, i will submit another pr for that

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/physical/PhysicalIcebergScanOperator.java (hunk 1 above)
Comment (dirtysalt) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/physical/PhysicalIcebergScanOperator.java:69]: We don't write `equal` method in `HDFSScanNodePredicates` class ?

**Review by dirtysalt — DISMISSED** [review]

**Comment by caneGuy** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 45 / 146 (30.82%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/IcebergScanImplementationRule.java|0|1|00.00%|[31]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|0|4|00.00%|[734, 776, 832, 833]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/ScanOperatorPredicates.java|0|7|00.00%|[68, 69, 70, 71, 72, 73, 74]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerTraceUtil.java|0|2|00.00%|[193, 426]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalIcebergScanOperator.java|0|10|00.00%|[30, 35, 40, 66, 67, 72, 78, 79, 80, 81]|
|:large_blue_circle:|com/starrocks/planner/IcebergScanNode.java|0|10|00.00%|[50, 136, 137, 190, 214, 216, 244, 246, 289, 295]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalHudiScanOperator.java|0|1|00.00%|[67]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalHiveScanOperator.java|0|1|00.00%|[67]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RemoteScanPartitionPruneRule.java|6|38|15.79%|[79, 80, 81, 82, 83, 84, 85, 86, 87, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 106, 107, 108, 110, 135, 136, 138, 139, 140, 141, 143, 144]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalIcebergScanOperator.java|3|6|50.00%|[49, 74, 86]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownMinMaxConjunctsRule.java|35|65|53.85%|[47, 48, 49, 73, 76, 77, 78, 80, 81, 83, 84, 86, 102, 103, 107, 108, 109, 110, 111, 112, 113, 115, 116, 119, 121, 122, 123, 124, 125, 126]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneHDFSScanColumnRule.java|1|1|100.00%|[]|

**Comment by caneGuy** [comment]
run starrocks_admit_test

**Review by dirtysalt — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/ScanOperatorPredicates.java [hunk 4]
```diff
@@ -61,4 +62,15 @@ public List<ScalarOperator> getMinMaxConjuncts() {
     public Map<ColumnRefOperator, Column> getMinMaxColumnRefMap() {
         return minMaxColumnRefMap;
     }
+
+    @Override
+    public boolean equals(Object o) {
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/ScanOperatorPredicates.java:67]: forget to override hashcode function？

Comment (caneGuy) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/operator/ScanOperatorPredicates.java:67]: done

**Comment by caneGuy** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 7 / 157 (04.46%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/IcebergScanImplementationRule.java|0|1|00.00%|[31]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|0|4|00.00%|[704, 746, 802, 803]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PruneHDFSScanColumnRule.java|0|1|00.00%|[116]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/ScanOperatorPredicates.java|0|8|00.00%|[68, 69, 70, 71, 72, 73, 74, 79]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OptimizerTraceUtil.java|0|12|00.00%|[188, 189, 190, 191, 192, 193, 397, 398, 399, 400, 401, 402]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalIcebergScanOperator.java|0|10|00.00%|[30, 35, 40, 66, 67, 72, 78, 79, 80, 81]|
|:large_blue_circle:|com/starrocks/planner/IcebergScanNode.java|0|10|00.00%|[50, 136, 137, 190, 214, 216, 244, 246, 289, 295]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalIcebergScanOperator.java|0|6|00.00%|[20, 49, 58, 63, 74, 86]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalHudiScanOperator.java|0|1|00.00%|[67]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalHiveScanOperator.java|0|1|00.00%|[67]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/RemoteScanPartitionPruneRule.java|1|38|02.63%|[63, 65, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 106, 107, 108, 110, 133, 134, 135, 136, 138, 139, 140, 141, 143, 144]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownMinMaxConjunctsRule.java|6|65|09.23%|[43, 46, 47, 48, 49, 50, 51, 56, 57, 58, 59, 61, 62, 69, 70, 71, 72, 73, 75, 76, 77, 78, 80, 81, 83, 84, 86, 92, 93, 94, 95, 96, 97, 98, 100, 102, 103, 106, 107, 108, 109, 110, 111, 112, 113, 115, 116, 119, 121, 122, 123, 124, 125, 126, 128, 134, 135, 136, 137]|

**Comment by caneGuy** [comment]
run starrocks_admit_test

**Review by Youngwb — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

**Pull request merged**